internal/controller/controllers/agent_controller.go

```diff
@@ -204,6 +209,95 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
 	return r.updateStatus(ctx, log, agent, &h.Host, h.ClusterID, nil, false)
 }
 
+func (r *AgentReconciler) shouldApproveMoreCSRs(node *corev1.Node) bool {
+	return !isNodeReady(node)
+}
+
+// Validate that the CSR can be approved
+func (r *AgentReconciler) shouldApproveCSR(csr *certificatesv1.CertificateSigningRequest, agent *aiv1beta1.Agent, validateNodeCsr nodeCsrValidator) (bool, error) {
+	x509CSR, err := getX509ParsedRequest(csr)
+	if err != nil {
+		return false, err
+	}
+	if !isCsrAssociatedWithAgent(x509CSR, agent) {
+		return false, nil
+	}
+
+	if !isCsrInTimeSpan(agent, csr) {
+		return false, errors.Errorf("CSR %s of agent %s/%s: creation timestamp %v is not in required time span",
+			csr.Name, agent.Namespace, agent.Name, csr.CreationTimestamp.Time)
+	}
+	return validateNodeCsr(agent, csr, x509CSR)
+}
+
+func (r *AgentReconciler) approveAIHostsCSRs(clients SpokeK8sClient, agent *aiv1beta1.Agent, validateNodeCsr nodeCsrValidator) {
+	csrList, err := clients.ListCsrs()
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to get CSRs for agent %s/%s", agent.Namespace, agent.Name)
+		return
+	}
+	for i := range csrList.Items {
+		csr := &csrList.Items[i]
+		if !isCsrApproved(csr) {
```

Comment:
If we filter for `Pending`, we don't need to check if it's `Approved`

Reply:
Same as above

Reply:
Putting it on hold